Openapi endpoint config tests

Azquelt opened the pull request:
For #25437

abutch3r committed:
Support updating of MP OpenAPI Document and UI Endpoint

Provide a common class that can handle both endpoints. Allows for validation of both endpoints and being able to respect the results of any failures.

Add a mpOpenAPI Server configuration element that can take values for either endpoint.

Update backend bundles to allow for custom endpoint allocation. Both for MP OpenAPI 1.0-1.1 (com.ibm.ws.microprofile.openapi) and MP OpenAPI 2.0-3.1+ (io.openliberty.microprofile.openapi.2.0.internal)

Add CustomUITest to Test UI Path updates

abutch3r committed:
Update to resolve comments

Resolve Modified behavior such that if the config is modified the WABs are unregistered and recreated with the config

Format classes

create separate messages for UI and document endpoints and use new enum to differentiate between which log message should apply

Azquelt committed:
MPOpenAPI: Fix project classpath settings

test and src directories missing from project classpath

Azquelt committed:
MPOpenAPI config code review changes

Azquelt committed:
OpenAPI UI config message review changes

Azquelt committed:
MP OpenAPI remove potential conflict warning

This was a warning to say that the configured value of an OpenAPI path
might conflict with another feature. We're not sure whether we need this
message and we're not happy with the current wording so we're taking
it out for now.

Azquelt committed:
MP OpenAPI config tests

Not complete yet

Azquelt commented:
Tests show a couple of issues:
1. The OpenAPI document servlet has a hard-coded check that `/openapi` is being requested
2. When a WAB context root is updated, the bundle containing it is restarted. In the case of the OpenAPI document servlet, it's in the same bundle as the `ApplicationRegistry`. Having that bundle restarted causes the `ApplicationRegistry` to forget which applications are running and the OpenAPI documents generated for them.

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,26 @@
+mpOpenAPI.desc=Configuration for MicroProfile OpenAPI.
+
+docPath=MicroProfile OpenAPI document endpoint path
+docPath.desc=Path to obtain OpenAPI 3.0/Swagger documents
```
do this line and line 26need a period, similar to line 20? Technically none of them are complete sentences, so you could remove the period at line 20.

Also wondering if there is any additional information about specifying the path that might help the developer- as is the descriptions just restate the information in the definition

Azquelt replied:
Looking at the existing docs, it looks like our exiting descriptions do end with a period, even when it's not a full sentence so I'll add one here.

I agree the descriptions should be more helpful, opened #25559

Changes will go into #25436

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
 OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W.explanation=The server is configured to create OpenAPI documentation for the application or module name, but no application or web module with that name was deployed.
 OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W.useraction=Check that all applications started correctly and that the names in the configuration property are specified correctly.
+
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I=CWWKO1668I: The OpenAPI UI path has been set to {0}.
```
```suggestion
OPEN_API_UI_PATH_UPDATE_CWWKO1668I=CWWKO1668I: The OpenAPI UI path is set to {0}.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
 OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W.useraction=Check that all applications started correctly and that the names in the configuration property are specified correctly.
+
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I=CWWKO1668I: The OpenAPI UI path has been set to {0}.
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I.explanation=The OpenAPI UI path has been set to a value that is not the default.
```
```suggestion
OPEN_API_UI_PATH_UPDATE_CWWKO1668I.explanation=The OpenAPI UI path is set to a value that is not the default.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I.explanation=The OpenAPI UI path has been set to a value that is not the default.
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I.useraction=No action is required.
+
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I=CWWKO1669I: The OpenAPI Document path has been set to {0}.
```
```suggestion
OPEN_API_DOC_PATH_UPDATE_CWWKO1669I=CWWKO1669I: The OpenAPI Document path is set to {0}.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_UPDATE_CWWKO1668I.useraction=No action is required.
+
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I=CWWKO1669I: The OpenAPI Document path has been set to {0}.
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.explanation=The OpenAPI Document path has been set to a value that is not the default.
```
```suggestion
OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.explanation=The OpenAPI Document path is set to a value that is not the default.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.explanation=The OpenAPI Document path has been set to a value that is not the default.
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.useraction=No action is required.
+
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E=CWWKO1670E: The OpenAPI UI path is invalid. The default value will be used instead. Check previous warnings for the cause of validation failure.
```
```suggestion
OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E=CWWKO1670E: The OpenAPI UI path is invalid. The default value is used instead. Check previous warnings for the cause of validation failure.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.useraction=No action is required.
+
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E=CWWKO1670E: The OpenAPI UI path is invalid. The default value will be used instead. Check previous warnings for the cause of validation failure.
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation=The OpenAPI UI path is invalid. The default value will be used instead.
```
```suggestion
OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation=The OpenAPI UI path is invalid. The default value is used instead.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation=The OpenAPI UI path is invalid. The default value will be used instead.
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.useraction=Check earlier error messages for the cause of validation failure.
+
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E=CWWKO1671E: The OpenAPI Document path is invalid. The default value will be used instead. Check previous warnings for the cause of validation failure.
```
```suggestion
OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E=CWWKO1671E: The OpenAPI Document path is invalid. The default value is used instead. Check previous warnings for the cause of validation failure.
```

dmuelle commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,35 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.useraction=Check earlier error messages for the cause of validation failure.
+
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E=CWWKO1671E: The OpenAPI Document path is invalid. The default value will be used instead. Check previous warnings for the cause of validation failure.
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.explanation=The OpenAPI Document path is invalid. The default value will be used instead.
```
```suggestion
OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.explanation=The OpenAPI Document path is invalid. The default value is used instead.
```

dmuelle reviewed (commented):
ID review- minor edits to avoid past perfect and future tenses in messages per IBM style

Azquelt commented:
Sorry, @dmuelle, this PR does not need messages reviewed.

The messages changes are from #25436 and will disappear from this PR once that one is merged. (We're having some problems getting it through a build at the moment).

I'd opened it as a draft and hadn't realised libby would tag you for review. I should have added the review not needed tags.

Azquelt commented:
I will apply @dmuelle's suggestions to #25436.

abutch3r committed:
Improve URL checking for OpenAPI document endpoint

remove hardcoding of /openapi for endpoint.

Azquelt committed:
Move MP OpenAPI servlets to separate bundles

We want to be able to change the context root of the MP OpenAPI servlet
at runtime. When we do this, the bundle serving the servlet gets
restarted.

This is fine for the servlets but not for the other components in the
same bundle.

Azquelt committed:
Add log message checks

Azquelt committed:
Minor fixes to the EndpointManager

abutch3r committed:
Add OpenAPI invalid config tests

Tests invalid path characters - special characters and `.` based paths
Conflicting paths between Doc and UI and with other applications

Parameterize the defaults for the two endpoints.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -0,0 +1,421 @@
+
+        server.startServer(false);
+
+        assertNotNull("Web Application fails to start due to context path conflict with OPENAPIUI bundle", "SRVE0164E.*OpenAPIUI");
```
Not actually checking the logs here.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -0,0 +1,421 @@
+
+        server.startServer(false);
+
+        assertNotNull("Document Web Appplication path contains invalid characters",
+                server.waitForStringInLog("CWWKO1676E.*$")); // check that error indicating that conflict is thrown
+        assertNotNull("Document Web Appplication path is invalid",
+                server.waitForStringInLog("CWWKO1671E.*$")); // check that error indicating that conflict is thrown
+        assertNotNull("UI Web Appplication path contains invalid characters",
+                server.waitForStringInLog("CWWKO1675E.*$")); // check that error indicates invalid characters is logged
+        assertNotNull("UI Web Appplication path is invalid",
+                server.waitForStringInLog("CWWKO1670E.*$")); // check that error indicating that a failure has occurred has been thrown
```
the `.*$` isn't necessary here

abutch3r committed:
fix up log statement tests

Add in example log entries relating to the expected App startup failures

Azquelt commented:
These changes have been rolled into https://github.com/OpenLiberty/open-liberty/pull/25436
